### PR TITLE
RSE-75: Fix Issues on View and Edit Period Forms

### PR DESCRIPTION
**Pull request by MiyaNoctem**
## Overview
Some problems were found when working with view/edit forms on the new membership periods entity.

## Before
View Membership: 
1. 'Estimated Legacy Period' should have a '?' in the field label. Should be 'Estimated Legacy Period ?'
2. Time value should not be displayed in Start Date & End Date fields.
3. Info message for 'Estimated Legacy Period' should be corrected as below.
4. 'Recurring Contribution' subheading should be 'Recurring Contributions'. Letter 's' is missing.
5. Can each Contribution have a tag as '(Recurring)' under each as displayed in Contribution tab view if its a recurring.
6. 'Delete' is missing for Completed contribution . Also 'More' or 3 dots option is missing for Pending contribution.
7. View Membership period heading should have 'Membership Type' & Term Details (Period 1, Period 2..) as given in mockup.
8. Updated Membership dates are reflected only after refreshing the page. Auto -refresh is not happening after Editing Membership dates.
9. Error message message content should be corrected to 
   "The new start date of this membership period overlaps with another activated membership period. please review your changes"
   instead of 
   "The new end date of this membership period overlaps with another activated membership period. please review your changes"
10. Whenever Edit is clicked, Edit Membership Period window is displayed with Calendar box of Start Date opened by default

## After
1. 'Estimated Legacy Period' now has a '?' in the field label.
1. Time value is not be displayed in Start Date & End Date fields.
1. Info message for 'Estimated Legacy Period' is corrected as per spec.
1. 'Recurring Contribution' was changed to 'Recurring Contributions'.
1. Replicated use of CiviCRM's functionality to load contributions in recurring contributions detail view, so now the "Recurring" label is added as normal.
1. Replicated use of CiviCRM's functionality to load contributions in recurring contributions detail view, so now the all actions are the same as on the contributions view.
1. View Membership period View title now has the format "[Membership Type] - Period [Term Number]".
1. Implemented auto-refresh when updating a period.
1. Error message message content was corrected to show if it is the start date or the end date which causes an overlap.
1. Whenever Edit is clicked, calendar for start date is not opened.

### CODE REVIEW

File: CRM/MembershipExtras/BAO/MembershipPeriod.php [hunk 1]
```diff
@@ -269,30 +266,66 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
...
     }
 
+    $overlappedMembershipPeriods = self::getOverlappingPeriods($periodID, $membershipID, $periodNewStartDate, $periodNewEndDate);
+    if ($overlappedMembershipPeriods->N > 0) {
+      $overlappingStartDate = new DateTime($periodNewStartDate);
+      $overlappingEndDate = new DateTime($periodNewEndDate);
+      $conflictingPeriodStartDate = new DateTime($overlappedMembershipPeriods->start_date);
+      $conflictingPeriodEndDate = new DateTime($overlappedMembershipPeriods->end_date);
+      
+      $isEndDateOverlap = $overlappingEndDate > $conflictingPeriodStartDate && $overlappingStartDate < $conflictingPeriodStartDate;
```
Comment (omarabuhussein) [inline CRM/MembershipExtras/BAO/MembershipPeriod.php]: regarding : 
```
 $isEndDateOverlap = $overlappingEndDate > $conflictingPeriodStartDate && $overlappingStartDate < $conflictingPeriodStartDate;
```

I think 

```
$isEndDateOverlap = $overlappingEndDate >= $conflictingPeriodStartDate
```

is enough

File: CRM/MembershipExtras/BAO/MembershipPeriod.php [hunk 2]
```diff
@@ -269,30 +266,66 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
...
+      if ($isEndDateOverlap) {
+        throw new CRM_Core_Exception(
+          'The new end date of this membership period overlaps with another
+          activated membership period. Please review your changes.'
+        );
+      }
+      
+      $isStartDateOverlap = $overlappingStartDate < $conflictingPeriodEndDate && $overlappingEndDate > $conflictingPeriodEndDate;
```
Comment (omarabuhussein) [inline CRM/MembershipExtras/BAO/MembershipPeriod.php]: also here, I think 
```
$isStartDateOverlap = $overlappingStartDate <= $conflictingPeriodEndDate
```
is enough

File: CRM/MembershipExtras/BAO/MembershipPeriod.php [hunk 3]
```diff
@@ -269,30 +266,66 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
...
+      if ($isStartDateOverlap) {
+        throw new CRM_Core_Exception(
+          'The new start date of this membership period overlaps with
+          another activated membership period. Please review your changes.'
+        );
+      }
+      
+      throw new CRM_Core_Exception('Undetermined problem found validating period overlapping.');
```
Comment (omarabuhussein) [inline CRM/MembershipExtras/BAO/MembershipPeriod.php]: I think 

`The new membership period dates are overlapping with another period dates, please validate the new dates`

is more clear

File: templates/CRM/MembershipExtras/Form/MembershipPeriod/EditPeriod.tpl [hunk 4]
```diff
@@ -1,9 +1,10 @@
 <div class="crm-block crm-form-block">
+  <input/>
```
[the remark this answers is not in the document]
Comment (MiyaNoctem) [reply templates/CRM/MembershipExtras/Form/MembershipPeriod/EditPeriod.tpl]: Absolutely none!

File: CRM/MembershipExtras/Page/MembershipPeriod/ViewPeriod.php [hunk 5]
```diff
@@ -1,24 +1,81 @@
...
+    $periodTermNumber = $this->getPeriodTermNumber($periodData['membership_id']);
+    $pageTitle = "View $periodMembershipType - Period $periodTermNumber";
+    
+    CRM_Utils_System::setTitle(E::ts($pageTitle));
+  }
+  
+  /**
+   * Obtains term number for the current period.
+   */
+  private function getPeriodTermNumber($membershipID) {
```
[the remark this answers is not in the document]
Comment (MiyaNoctem) [reply CRM/MembershipExtras/Page/MembershipPeriod/ViewPeriod.php]: Done.

File: CRM/MembershipExtras/BAO/MembershipPeriod.php [hunk 6]
```diff
@@ -269,30 +266,66 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
...
     }
 
     if (empty($periodNewStartDate)) {
       throw new CRM_Core_Exception("Can't calculate period overlapping without period start date!");
     }
 
+    $overlappedMembershipPeriods = self::getOverlappingPeriods($periodID, $membershipID, $periodNewStartDate, $periodNewEndDate);
+    if ($overlappedMembershipPeriods->N > 0) {
+      $overlappingStartDate = new DateTime($periodNewStartDate);
+      $overlappingEndDate = new DateTime($periodNewEndDate);
```
[the remark this answers is not in the document]
Comment (MiyaNoctem) [reply CRM/MembershipExtras/BAO/MembershipPeriod.php]: Done, and refactored the logic into its own function, to clean-up the code.

File: CRM/MembershipExtras/BAO/MembershipPeriod.php [hunk 7]
```diff
@@ -269,30 +266,66 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
...
+      $conflictingPeriodStartDate = new DateTime($overlappedMembershipPeriods->start_date);
+      $conflictingPeriodEndDate = new DateTime($overlappedMembershipPeriods->end_date);
```
[the remark this answers is not in the document]
Comment (MiyaNoctem) [reply CRM/MembershipExtras/BAO/MembershipPeriod.php]: Done and refactored.

File: CRM/MembershipExtras/BAO/MembershipPeriod.php [hunk 8]
```diff
@@ -269,30 +266,66 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
...
+   * @param string $periodNewStartDate
+   * @param string $periodNewEndDate
+   * 
+   * @return \self
+   */
+  private static function getOverlappingPeriods($periodID, $membershipID, $periodNewStartDate, $periodNewEndDate) {
     $overlappedMembershipPeriods = new self();
     $overlappedMembershipPeriods->membership_id = $membershipID;
     $overlappedMembershipPeriods->is_active = TRUE;
+    $overlappedMembershipPeriods->whereAdd('id != ' . $periodID);
```
[the remark this answers is not in the document]
Comment (MiyaNoctem) [reply CRM/MembershipExtras/BAO/MembershipPeriod.php]: I'm going to bring in all periods instead, and use the new methods to check if either start or end date overlaps for each.

File: CRM/MembershipExtras/BAO/MembershipPeriod.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (MiyaNoctem) [reply CRM/MembershipExtras/BAO/MembershipPeriod.php]: Done. Now errors are stored in an array and a single exception is thrown.

Comment (MiyaNoctem) [reply CRM/MembershipExtras/BAO/MembershipPeriod.php]: I left checking of overlapping as generic as possible, so we can use it later on other parts if we need to. So basically, a start date overlaps if it is between the start and end dates of another period. And and end date overlaps if it is within the start and end dates of another period. Taking into account periods can have an empty end date.

Comment (jamienovick) [reply CRM/MembershipExtras/BAO/MembershipPeriod.php]: I could be wrong but a period could still overlap even if the start or end date is not within the start dates of another period if it starts before and ends after another period. Just FYI.

Comment (omarabuhussein) [reply CRM/MembershipExtras/BAO/MembershipPeriod.php]: make sense @MiyaNoctem  but just a small correction regarding
> a start date overlaps if it is between the start and end dates of another period

this should not be just  **between**, but rather **between or equal**


@jamienovick that's correct and we already talked about it in another comment above.

File: CRM/MembershipExtras/BAO/MembershipPeriod.php [hunk 9]
```diff
@@ -269,28 +265,94 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
...
+    }
+
+    $endDate = new DateTime($periodEndDate);
+    if (empty($evaluatedPeriod->end_date)) {
+      return $endDate >= $evaluatedPeriodStartDate;
+    }
+
+    $evaluatedPeriodEndDate = new DateTime($evaluatedPeriod->end_date);
+
+    return $endDate >= $evaluatedPeriodStartDate && $endDate < $evaluatedPeriodEndDate;
```
Comment (omarabuhussein) [inline CRM/MembershipExtras/BAO/MembershipPeriod.php]: shouldn't 

`$endDate < $evaluatedPeriodEndDate;`

be

`$endDate <= $evaluatedPeriodEndDate;`

**Pull request merged**